### PR TITLE
Reactivate atoms and bonds valence checks

### PR DESCRIPTION
**Pull request by andrrizzi**
This partially re-implement issue #185 . It reimplements valence checks for atoms and bonds only.

Reactivating valence checks for angles and torsions will require larger modifications, but I'd like to take a decision about #216 first in order to minimize the total time I'll spend on this, and implement it in a separate PR.

### CODE REVIEW

**Comment by andrrizzi** [comment]
I've also renamed `ParameterHandler.get_matches()` to `find_matches()` for consistency with `ToolkitWrapper.find_smarts_matches()` because I thought it was more representative of what it is actually doing, but I can revert that commit if you don't agree.

**Comment by j-wags** [comment]
This pull request **fixes 1 alert** when merging 6fbe300b40d8187e41b245a7323c24ce8afc406e into 8ea7a09e9fdba58a0e2567da7be37e94a675aa7c - [view on LGTM.com](https://lgtm.com/projects/g/openforcefield/openforcefield/rev/pr-8e20ee96d4d296ae3e5bebb3aa1e8b907be668df)

**fixed alerts:**

* 1 for Unused import

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: openforcefield/typing/engines/smirnoff/parameters.py [hunk 1]
```diff
@@ -825,6 +849,58 @@ def to_dict(self, output_units=None, return_cosmetic_attributes=False):
...
+                        "- {unassigned_str}").format(parameter_handler=cls.__name__,
+                                                     unassigned_str='\n- '.join(unassigned_terms))
+        if len(not_found_terms) > 0:
+            if err_msg != "":
+                err_msg += '\n'
+            err_msg += ("{parameter_handler} assigned terms that were not found in the topology:\n"
+                        "- {not_found_str}").format(parameter_handler=cls.__name__,
+                                                    not_found_str='\n- '.join(not_found_terms))
+        if err_msg != "":
+            raise RuntimeError(err_msg)
```
Comment (jchodera) [inline openforcefield/typing/engines/smirnoff/parameters.py]: I don't think we want to raise `RuntimeError`, since these are not going to be informative about what was not matched. Instead, why don't we define our own exception hierarchy (e.g. `UnassignedParameterException > UnassignedValenceParameterException > UnassignedBondParameterException`) so that applications or tests could catch these exceptions and proceed with execution if intended. (For example, someone might be mixing a force field and NN potential outside of SMIRNOFF.)

Also, let's be sure the docstrings document exceptions we intentionally throw under different circumstances.

Comment (andrrizzi) [reply openforcefield/typing/engines/smirnoff/parameters.py]: Will do, thanks!

**Comment by andrrizzi** [comment]
After conversation in #216, I've decided to implement `Topology.angles/propers/impropers` properties for now returning tuples of `Atom`s, similar to what's currently implemented in `Molecule`. This check will be important to include in the release. In the future, we may want to have `Angle/Proper/Improper` classes similar to `Bond` to make things a little more consistent.

**Comment by jchodera** [comment]
Please remember to add a note to the docstring for any methods or properties where the API may not be final!

**Comment by andrrizzi** [comment]
@j-wags can correct me if I'm wrong, but I'm not sure we're at that level yet, meaning that I think we're planning to generally label the API as unstable for the upcoming release.

I think it's still hard to make guarantees about what's not going to change signature in the next versions except maybe for a few methods such as `ForceField.create_openmm_system()` and `Molecule.from_anything()`, which are needed for the most common use cases of the toolkit. We're planning to take a systematic pass at the API in the short future though.

**Comment by andrrizzi** [comment]
Reactivating the valence terms caused some tests to fail. Most of them complain about improper torsions that are not parametrized, but also some proper torsions as well. I'll check a few specific cases to see if this is a bug or something that we want to ignore.

**Comment by j-wags** [comment]
This pull request **introduces 6 alerts** and **fixes 1** when merging 426e345e0843623a3f54082cc5696782eb99325c into 8ea7a09e9fdba58a0e2567da7be37e94a675aa7c - [view on LGTM.com](https://lgtm.com/projects/g/openforcefield/openforcefield/rev/pr-4b77780cc32ef30be2dfcf2ab0eed6686992de73)

**new alerts:**

* 6 for Property in old\-style class

**fixed alerts:**

* 1 for Unused import

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by j-wags** [comment]
> I think we're planning to generally label the API as unstable for the upcoming release.

Right. I get the sense that we're applying "unstable API" warnings to 80% of functions right now, with no clear pathway to un-labeling them as such. If we define a process for removing the label, then it makes more sense.

Before I started typing this, I had probably agreed with @andrrizzi -- Everything will be labeled as "unstable", and then the day we do the stable release, all the labels are removed. However, maybe it would make more sense to have a meeting after the initial release, and select different chunks of the API to "stabilize" as part of different milestones. This way, we could have focused discussions about individual areas of the API and remove the "unstable" labels one at a time.

tl;dr - Let's apply "unstable API" labels to individual functions. Later, we can come through and remove them one-by-one as we "stabilize" different areas.

**Comment by davidlmobley** [comment]
Impropers not parameterized is OK -- impropers are optional, essentially. But propers not parameterized is concerning.

**Comment by j-wags** [comment]
This pull request **introduces 6 alerts** and **fixes 1** when merging 20ea3c02ec3c3df2ad75ddec6ac0ad4f8ad2860e into 8ea7a09e9fdba58a0e2567da7be37e94a675aa7c - [view on LGTM.com](https://lgtm.com/projects/g/openforcefield/openforcefield/rev/pr-fa87eb0d2e3fb74564a580d3011bf0cc412d5ebd)

**new alerts:**

* 6 for Property in old\-style class

**fixed alerts:**

* 1 for Unused import

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by j-wags** [comment]
This pull request **introduces 6 alerts** and **fixes 1** when merging a21ae75248c1fd042c22faac8f3e32584d7c0498 into 8ea7a09e9fdba58a0e2567da7be37e94a675aa7c - [view on LGTM.com](https://lgtm.com/projects/g/openforcefield/openforcefield/rev/pr-a38a64951746f4589c01973fb9be96e095ec6dda)

**new alerts:**

* 6 for Property in old\-style class

**fixed alerts:**

* 1 for Unused import

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by andrrizzi** [comment]
Thanks for the fix, @j-wags! Good to know that our tests caught that. This is ready for merging on my end.

**Comment by j-wags** [comment]
This pull request **introduces 6 alerts** and **fixes 1** when merging 826716bf21f7716414834cc181726b8787b7928c into 8ea7a09e9fdba58a0e2567da7be37e94a675aa7c - [view on LGTM.com](https://lgtm.com/projects/g/openforcefield/openforcefield/rev/pr-6939c20d3c85c391f95e6f22ad5ddc4445ac46eb)

**new alerts:**

* 6 for Property in old\-style class

**fixed alerts:**

* 1 for Unused import

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by jchodera** [comment]
If we are not certain that most of the API will be stable at release, we really should NOT be calling this a 1.0 alpha release, since it's absurd to have to go through some crazy gymnastics of a few dozen alpha bugfix, feature, and API breaking releases and then switch to semantic versioning. Let's stick with 0.y.z releases until we feel that much of the API is stable, and not worry about whether or not the April release is "1.0 anything". It's just the first release with RDKit support and a more coherent object model for the parameter set, but it can be 0.y.z. Feature and API breaking releases can increment y, and bugfixes can increment z. If we release 1.0.0 in year one, I think we will still be in great shape.

In addition, we want to label the docstrings of classes and functions that have unstable APIs with standard warnings. We can remove those as parts of the API stabilize.

TensorFlow has been in a similar situation for a while, and we are a much smaller operation, so there is nothing to be ashamed about here. We just need to be very clear in how we communicate this to our users

**Comment by j-wags** [comment]
I agree with @jchodera that the upcoming release should be 0.2.0, and that we can build toward 1.0.0. It's OK if the majority of the API is labeled as unstable for 0.2.0, but we should have a plan to collect API feedback after the initial release, and have a schedule that will have us finalizing it in manageable-sized chunks.

File: openforcefield/typing/engines/smirnoff/parameters.py [hunk 2]
```diff
@@ -1387,22 +1476,21 @@ def create_force(self, system, topology, **kwargs):
...
         # Create all particles.
         for particle in topology.topology_particles:
             force.addParticle(0.0, 1.0, 0.0)
 
         # Set the particle Lennard-Jones terms.
-        for (atoms, ljtype) in atoms.items():
-            force.setParticleParameters(atoms[0], 0.0, ljtype.sigma,
+        for atom_key, ljtype in atoms.items():
+            atom_idx = atom_key[0]
+            force.setParticleParameters(atom_idx, 0.0, ljtype.sigma,
```
Comment (j-wags) [inline openforcefield/typing/engines/smirnoff/parameters.py]: Nice. This code is much clearer.

**Comment by jchodera** [comment]
Sounds like a great plan, @j-wags.

We should also be sure to craft a release announcement for the News page that lays out some details of where we are, what feedback we'd like (and how), and where we're going (and how to check the Milestones).

**Comment by andrrizzi** [comment]
>  we really should NOT be calling this a 1.0 alpha release

Absolutely agree.

Thank you all for the reviews!
